### PR TITLE
RVF tuples of arrays and array views

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR makes tuples of arrays and views RVFable if all the elements are
serializable.

The motivation for this is to be able to RVF tuples containing slices, so that
we can turn `chpl_serializeSlices` on by default. We know that it helps in many
cases, but it hurts in some.

(This PR doesn't help promotion with slices. So, we are still keeping that flag off) 

Follow up to https://github.com/chapel-lang/chapel/pull/12985

Resolves https://github.com/chapel-lang/chapel/issues/14530
Resolves https://github.com/Cray/chapel-private/issues/226
Resolves https://github.com/Cray/chapel-private/issues/1461

### Motivation

Consider

```chapel
forall (x,y) in zip(myArr, mySlice) do x=y;
```

The follower loop that's inside the `on` statement needs to access `myArr` and
`mySlice`. We privatize the former, so accessing its instance is free of
communication. However, we don't want to privatize `mySlice` to keep creating
slices as light-weight as possible. This creates a weird situation where we RVF
the view's record, and know that the data that it refers to is local, however,
the instance remains in the originating locale. To be able to forward that
instance, we need the `chpl__iterLF` tuple we create to be forwardable.

This `forall` loop is also what we currently get with assignments involving
slices.

When `chpl_serializeSlices=true` we see some GETs for the snippet above, with
this branch, we do not see any.

### Implementation Details

- The implementation relies on some type variables that store ArrayView types.
  As they are also wrapped in `record _array`, they are actually runtime types.
  However, the compiler wasn't setting up the `runtimeTypeMap` accordingly. So,
  this PR adds `mapArrayViewsToRuntimeTypes` function that is called when we are
  resolving a type specifier. It maps an AggregateType that refers to an
  ArrayView into the runtime type of the original array.

- We need to be able to return a tuple of ArrayViews without copying them,
  because such copies result in creating a new array. To that end, this PR adds
  `no copy return` to tuple deserializer. But I have realized that that pragma
  wasn't being taken into account for functions that returns tuples. So, I also
  adjust that in the compiler.

- We are not copying when we return, and this results in returning a ref tuple.
  (such tuples are not actually ref tuples, because of weird situation of
  ArrayViews that are in between `ref` and value). We want the instances created
  within the tuple deserializer to *not* be freed by it but by its caller. In
  order to achieve this:

  - In `callDestructors` we thwart cleaning up any locals that are in tuple
    deserializer. (Probably the ugliest part of this PR)

  - And then, when we are in `fixUpDestructors` we create a special body for the
    tuple deinitializer if it contains such data that needs to be freed. This
    allows us to cleanup the memory after deserialized ArrayView instances
    stored in tuples. Without this, the tuple deinitializer gets an empty body.

- We do not serialize tuples whose items are all privatized types. By definition
  a privatized instance should have a replica instance on every locale.
  Therefore, RVFing another one is redundant.

- We do not serialize privatized arrays if "they are by themselves", but we do
  serialize them if they appear in a tuple. The reason for that is similar to
  the above: we don't want to serialize-and-forward something that is already
  privatized. Morever, doing that breaks fast-on optimization because
  `chpl__deserialize` is not a "fast" function. This is achieved by making
  `_array.chpl__serialize` not resolve if the `_instance` is privatized and
  adding a `_array.chpl__forceSerialize` that doesn't check for privatization.
  The latter is used within the tuple serialization logic.

### Test status

- [x] gasnet `release/examples`
- [x] gasnet `release/examples` --memleaks
- [x] gasnet.quickstart `release/examples` with asan
- [x] gasnet `distributions`
- [x] gasnet `optimizations`
- [ ] standard
- [ ] gasnet
- [ ] ugni `distributions` to get the commcounts tests right

### Post-merge notes

- We still need to apply a similar fix for promotion with slices. After that we can
  turn `chpl_serializeSlices` on.

- If we had a better story for ref fields, it might have helped implementing
  this. Currently, its lack, and ArrayViews being somewhere between a ref and a
  value causes some confusion. For example, we have `var _instance` in `record
  _array`, but an ArrayView instance cannot technically be `var`.

  - This shows up in cases with the compiler's typedef for arrayview tuples. I
    see things like:

    ```c
    typedef
    _array_ArrayViewSliceArr_int64_t_int64_t_BlockDom_1_int64_t_F_DefaultDist_int64_t_BlockArr_1_int64_t_F_int64_t_DefaultDist_chpl_bool_chpl
    _tuple_2_star__ref__array_ArrayViewSliceArr_int64_t_int64_t_BlockDom_1_int64_t_F_DefaultDist_int64_t_BlockArr_1_int64_t_F_int64_t_DefaultDist_chpl_bool_chpl[2];
    ```

    which is confusing, because the actual type is a static array of
    `_array(Slice)[2]`, but we typedef it to `2*ref(_array(Slice))`.

- Probably related, but compiler's general handling of ArrayView can also be
  confusing. I was never quite sure when to call `getValType` on an
  AggregateType to get the correct behavior.

- Privatization and serialization plays an interesting role on arrays. We don't
  really want to serialize-and-forward a privatized instance, because (1) it is
  redundant, (2) deserialization inside the `on` function thwarts fast on
  optimization. However, we want to serialize-and-forward tuples containing
  non-privatized instances even if they also contain privatized instances. More
  than that, when we serialize we just put their `pid` in the serialized buffer,
  so we technically rely on privatization while serializing arrays. I think it
  is an interesting problem to think about creating an optimization ecosystem
  where these two can coexist more comfortably/naturally :)

- The way we figure out whether something is serializeable or not relies too
  much on being able to resolve `chpl__serialize` (and not as much
  `chpl__deserialize`). I hope that especially with constrained generics, we
  can have a more principled story for doing this. Without constrained generics,
  the way we check for privatization via `dsiSupportsPrivatization` feels much
  better. (I hope that we can make both optimizations rely on constrained
  generics in some form)

- Somewhat related: I think the serialization interface should support two
  different types of serialization. (1) serialization by itself, (2)
  serialization along with other serialized items. In other words, there might
  be cases where we don't really want to serialize-and-forward something, but it
  might make sense to do that if it is a field in a record or a tuple, for
  example. There's no clean way to express that at the moment. (I attempted this
  by adding `chpl__forceSerialize`

### CODE REVIEW

**Comment by e-kayrakli** [comment]
### Some performance improvements (16 locale XC)

- @npadmana's [FT](https://chapel-lang.org/perf/16-node-xc/?startdate=2019/06/06&enddate=2020/12/14&configs=gnuugniqthreads&graphs=npbftperfmflopsssized)

This test threw in `chpl_serializeSlices`, but it will be on-by-default and not
needed to be thrown in by the user anymore. This helps the case without
primitive optimization, because that case uses direct GET/PUT instead of a slice
assignment.

|                       | master w/ flag (GFLOPS) | master w/o flag | branch
|-----------------------|-------------------------|-----------------|-------
|w/  primitive comm     | 403                     | 403             | 404
|**w/o primitive comm** | **261**                 | **25**          | **260**


- [BlockDist transfer commcounts](https://chapel-lang.org/perf/16-node-xc/?startdate=2020/08/28&enddate=2020/12/14&configs=gnuugniqthreads&graphs=blockdisttransfern1e6)

With slice serialization on, we see the following change in `on` statements,
without any noticeable change in GETs/PUTs

|          | master |  branch
|----------|--------|--------
|elegant   | ~600   |  ~420
|DestLocal | ~320   |  ~230


- [BlockDist transfer times](https://chapel-lang.org/perf/16-node-xc/?startdate=2020/08/28&enddate=2020/12/14&configs=gnuugniqthreads&graphs=blockdisttransfern1e6)

This is a noisy test and I haven't averaged but roughly I think we'll get som
improvement in `elegant` and `DestLocal` cases

|          | master (ms) | branch (ms)
|----------|-------------|------------
|elegant   | 165-191     | 150-181
|DestLocal | 135-150     | 125-143

- There are also some commcount changes in the tests.

File: test/optimizations/tupleSerialization/sliceCombos.chpl [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+forall (x,y) in zip(A[innerDom], B) do x=y;
+endDiag("zip(slice, arr)");
+
+startDiag();
+forall (x,y) in zip(B, A[innerDom]) do x=y;
+endDiag("zip(arr, slice)");
+
+startDiag();
+forall (x,y) in zip(A[innerDom], C[innerDom]) do x=y;
+endDiag("zip(slice, slice)");
```
Comment (e-kayrakli) [inline test/optimizations/tupleSerialization/sliceCombos.chpl:24]: I'll add tests loops with larger number of iterands

**Comment by e-kayrakli** [comment]
There are some challenges with merging this PR that I wanted to note them before I forget. 

- `chpl__iterLF` is a ref tuple when you are zippering two `_array`s. So, in deserializing that tuple, one would expect to get another ref tuple. Because a deserializer should always return an object of the same type as the original one. 
- Based on that, it is not quite straightforward how you would achieve that and get the performance benefits you expect. When serializing a ref tuple, we are supposed to serialize the instances that are referenced. And on the deserializing end, we need to create new instances and return a tuple referencing those new instances. But then where do you store those new instances? This PR creates them as local variables in `chpl__deserialize` and creates a backdoor in the compiler that doesn't clean memory for the local variables in that function. Those instances are cleaned up in the tuple deinitializer, instead. Not ideal.
- A solution to this can be by deserializing into a local temp inside the on function, and creating a ref temp to that value temp and using it as the actual item that we needed. This goes against the intuition in the first bullet. (I see that we have a value and a ref version of `chpl__iterLF` at some points in the AST, but I can't tell why. Maybe it is because of something similar?) Doing this may also require a separate way for cleaning the deserialization data that we may require the type implementer to write similar to `chpl__serialize` and `chpl__deserialize` (or whatever we end up calling them when we make them more user-facing). I am not quite sure how much I like this idea yet.
- On top of this, there is also a confusion with the tuples of ref arrays. Although we seemingly can have such things in the AST, in the generated code I see such types as typedefs to value-like types. This is slightly less of a problem with star tuples as "ref" and "value" versions typedef to the same static array type. However, with heterogeneous tuples, they typedef to different struct types, although those structs have the identical fields.

**Comment by bradcray** [comment]
Capturing a notion I brought up in discussion today before I forget:

Today, when doing serialize/deserialize, I think the compiler stamps out code according to the following template:

```chapel
var state = something.chpl_serialize();
on remoteLocale {
  var localSomething = something.type.chpl_deserialize(state);  // as Engin says in point 1, `localSomething.type == something.type`, which seems attractive to preserve
  // ...compute with `localSomething` rather than `something`
}
```

My question was, could we modify this to something like the following:

```chapel
var state = something.chpl_serialize();
on remoteLocale {
  var localSomething = something.type.chpl_deserialize(state);  // as Engin says in point 1, `localSomething.type == something.type`, which seems attractive to preserve
  // ...compute with `localSomething` rather than `something`
  localSomething.chpl_cleanupDeserializedData();
}
```

where someone creating serialization for a given type would provide three functions (serialize, deserialize, cleanup) rather than just two as today.  My thoughts in suggesting this are:

1) preserves Engin's first bullet which resonates with me
2) possibly moves complexity of whether or not we have to clean something up from the compiler to module code?

That said, my other thought is what the downsides would be of having `chpl_deserialize()` return a different type than the one on which it's defined...  I like the symmetry of preserving the type, but... how necessary is it? / what would break or be worse if it did?

**Comment by e-kayrakli** [comment]
@bradcray -- Thanks for outlining this idea. I have alluded to it in the 3rd bullet, but not that much in detail.

I have mentioned this in the meeting but, I am not a big fan of having a third method for cleanup. In most cases I'd expect the required cleanup to be identical to what `deinit` is doing for the given type that is returned by `chpl__deserialize`. If we were to have those 3 methods in our "serialization interface", can I omit the cleanup function and expect my regular `deinit` method to be called, or is it a resolution error to omit that?

This PR shows that there may be some exceptional cases where regular deinit may not be enough. In those cases, I am thinking the type can store whether a given instance is a result of deserialization in a field and do different things in its `deinit`. Initially, I was thinking a const bool field that can take care of that, but I didn't like that it would increase a tuple's size (or any other serializable type for that matter). I am wondering whether we can have a param bool, and different instantiations of tuple type be coercible if that field is the only difference in types. (A very similar idea of having a param field that's inert to resolution came up in const domain work where I have toyed with the idea of having a `param definedConst: bool` field that would generate coercible instantiations, if it is the only difference between two types)

**Comment by bradcray** [comment]
> In most cases I'd expect the required cleanup to be identical to what deinit is doing for the given type that is returned by chpl__deserialize. If we were to have those 3 methods in our "serialization interface", can I omit the cleanup function and expect my regular deinit method to be called, or is it a resolution error to omit that?

I agree that they'd typically be the same.  I think it seems attractive to make supplying the cleanup method optional and to just do normal deinit if it's not provided (or to have the compiler generate a cleanup routine that calls deinit).

> but I didn't like that it would increase a tuple's size

I agree with that.

> I am wondering whether we can have a param bool, and different instantiations of tuple type be coercible if that field is the only difference in types. 

It seems as though if we do that, though, we're arguably breaking the "deserialized type is the same as the serialized type" bullet, though.  Which is part of what made me wonder "How bad is that really?"  It's a principle that I agree with based on my (very little amount of) previous work with serialize/deserialize, but maybe there really isn't any reason to keep them the same.  E.g., if the `on-clause` body is generated to deal with the deserialized type, does it really matter whether or not it matches the original type?  If not, I'd be inclined to have deserialize create a value tuple that cleans itself up rather than adding more special cases to tuples ("These types aren't really the same, but we'll pretend they are").

**Comment by e-kayrakli** [comment]
@bradcray -- is it too late to back away from this? :) Over the break, I was thinking a bit more about this, and my feet got even colder.

- The tuple we create for a zip clause is a ref tuple. That makes sense. At least, ref is an obvious choice and I am not sure what would it mean to make that a value tuple (wrapper record values that reference the same instance as the original?)
- This is very fundamental, but I don't remember discussing it; does it even make sense to serialize a ref tuple or any ref variable? Shouldn't we expect it to create another ref that references the same original value?

So, I feel like these two are what has been causing all the trouble implementing this, and I feel like we should find alternative ways to implement this optimization.

What I am thinking for this is "lazy privatization". We can RVF the privatization data for these instances and privatize the instances within the `on` wrapper just before executing the `on` function. To make the memory management more principled we can think of privatization contexts for on statements, where some instances are privatized only during an on statement. I think an implementation for that can also form the basis for privatization 2.0.

What do you think?

**Comment by bradcray** [comment]
> is it too late to back away from this?

Of course not!

> The tuple we create for a zip clause

When I think about big/brave changes we could make to make things better, one of the things I wonder is whether we should take this opportunity to move away from representing zip clauses as tuples.  It doesn't make the problem completely go away, but it gets tuples out of the mix, and becomes more a question of serializing a varargs list of expressions which somehow seems more appealing.

> does it even make sense to serialize a ref tuple or any ref variable? Shouldn't we expect it to create another ref that references the same original value?

I would expect serialization of a ref to result in a wide (cross-locale) ref back to the original variable.  I think we discussed this at one point, but it does have the potential complication that it could result in a type change (because a local ref, once deserialized, would become a wide ref).

> What I am thinking for this is "lazy privatization".

I don't understand this well enough yet to have an opinion about it.

**Comment by e-kayrakli** [comment]
> I would expect serialization of a ref to result in a wide (cross-locale) ref back to the original variable. I think we discussed this at one point, but it does have the potential complication that it could result in a type change (because a local ref, once deserialized, would become a wide ref).

That's what I'd expect, too. But doing it that way defeats the purpose of the optimization, because accessing the instance is still going to be with communication.

And doing it the way we do it in the PR is not proper serialization to me.

**Comment by bradcray** [comment]
> But doing it that way defeats the purpose of the optimization

Right, I think this came up before because we were asking the question "What if the things being zippered include something we want to serialize (e.g., a block array or slice) and something that we traditionally haven't been able to?"  If our only choices were to serialize the whole tuple or nothing, my proposal was to serialize the whole thing where, for any field that didn't support serialization, we'd just deserialize as a reference back to the original thing (since that's no worse than how it'd be treated otherwise, while not serializing/deserializing the block array or slice seems like a performance killer).

**Pull request closed without merging**